### ISSUE TITLE
Feature idea: proper API-driven deep copy of AWX objects

### ISSUE THREAD

**Issue by wenottingham** [ansible/awx#283]
##### ISSUE TYPE
 - Feature Idea

##### COMPONENT NAME
 - API
 - UI

##### SUMMARY

Assorted items in AWX have a 'copy' button in the UI to create a duplicate item that can be edited.  Notably, the job template implementation of this is entirely UI driven, which means that it doesn't copy related items (surveys, labels, and more) reliably.

To fix this, all relevant items should have API-based implementations, that properly copy related items, and the UI should be updated to use it.

##### ADDITIONAL INFORMATION

Should be done for:

- [ ] job templates
- [ ] projects
- [ ] inventories
- [ ] workflows
- [ ] credentials
- [ ] notifications
- [ ] inventory scripts
- [ ] schedules?
- [ ] teams?

**Comment by matburt** [comment]
This can almost certainly be generalized on the views while utilizing the serializer to grab the required fields. `CommonModelName` usage can help us generate a named based on the original. The current UI copy fields generate an iso8601 addition to the name... not sure how I feel about that one going forward with all of these though.

**Comment by AlanCoding** [comment]
My big question here is whether we should continue using a copy sub-endpoint like we have no with `/workflow_job_templates/N/copy/`, or if we should use the http COPY method on the object's endpoint. That is, submitting a COPY request to `/job_templates/N/`, as opposed to a POST to `/job_templates/N/copy/`.

EDIT, ANSWER: we will continue to use the copy sub-endpoint.

**Comment by AlanCoding** [comment]
> notifications

I assume this means Notification Templates

**Comment by AlanCoding** [comment]
Note from feature planning:

Most of the really hard questions about the RBAC implications of this are tabled for the planning of #166, which will be changing those things anyway.

**Comment by jangsutsr** [comment]
Planning meeting summary:
- [x] Remove uuid
- [x] Give user the ability to specify the name of copied object in POST body
- [x] For deep copy, return the top level object and populate its related objects later in celery task. Find a good return status code.
- [x] Use `can_create` for copy permission check for now, modify when implementing granular permissions.
- [x] Deal with workflow corner case by recursively check node permissions in GET request-response cycle.
- [x] Do not copy inventory source linkage in inventory copy.
- [x] Do not copy related roles, create admin role of copier for the copied object.
- [ ] Websocket in copy task -- table for later.

**Comment by AlanCoding** [comment]
> For deep copy, return the top level object and populate its related objects later in celery task. Find a good return status code.

This is tricky, because the object was created, but operations are still on-going. I would still favor a 202 code. @matburt yay or nay?

Rest of that looks great. Good job!

**Comment by matburt** [comment]
`201` is better than `202` because we are making a commitment to creating the primary object. With 202 I'd suspect a chance of not creating any objects at all which is not the case.

**Comment by AlanCoding** [comment]
@jakemcdermott is this clearly on the radar of someone on the UI team? I just hit a bit of fallout from the UI lagging the API in this case:

 - Create a JT with a survey with password-type questions
 - copy that JT in the UI

This results in a bad response code from the API and in the UI

> Call failed. Return status: undefined

That bug will only be fixed by the UI switching to this new API-driven deep copy mechanism.

**Comment by jakemcdermott** [comment]
@AlanCoding 
Where did we land on RBAC / how clients can know whether or not they have the permission to copy something? This is still needed, I believe.

**Comment by jakemcdermott** [comment]
I don't think the API implementation for copying has landed yet.. You might be encountering a different bug?

**Comment by wenottingham** [comment]
@jakemcdermott has this all landed (i.e., can it be closed)

**Comment by jakemcdermott** [comment]
@wenottingham 
We're close to that point - there's a PR in the works api side for https://github.com/ansible/awx/issues/1167. Once that's in it's a very quick update UI side. I'll close this issue once these things are wrapped up.

**Comment by AlanCoding** [comment]
Yeah, I'm slogging through this right now. The trouble isn't providing the field, that's basically trivial. But doing it _efficiently_ and elegantly is still a relatively small problem, but it's very difficult to think through.

I might have to just put up the inefficient version for Jake first, because I'm getting bogged down by the internal piping here.

**Comment by jian-feng** [comment]
I hope this one could be done soon. 
Our customer got stuck in #383, which is pending on this.

Best regards,

**Comment by AlanCoding** [comment]
The part of it relevant to that should already be done.

**Comment by matburt** [comment]
If they are a Tower customer then this will be available in the next Tower release which is not ready yet @jian-feng

**Comment by jakemcdermott** [comment]
This has been handled downstream and will be resolved in AWX when the associated code changes are merged into AWX.

**Comment by wenottingham** [comment]
@AlanCoding can this be closed?

**Comment by AlanCoding** [comment]
yes
